libsolidity/formal/VariableUsage.cpp

```diff
@@ -19,72 +19,61 @@
 
 #include <libsolidity/formal/SMTChecker.h>
 
-#include <libsolidity/ast/ASTVisitor.h>
+#include <algorithm>
 
 using namespace std;
 using namespace dev;
 using namespace dev::solidity;
 
-VariableUsage::VariableUsage(ASTNode const& _node)
+void VariableUsage::endVisit(Identifier const& _identifier)
 {
-	auto nodeFun = [&](ASTNode const& n) -> bool
-	{
-		if (auto identifier = dynamic_cast<Identifier const*>(&n))
-		{
-			Declaration const* declaration = identifier->annotation().referencedDeclaration;
-			solAssert(declaration, "");
-			if (VariableDeclaration const* varDecl = dynamic_cast<VariableDeclaration const*>(declaration))
-				if (identifier->annotation().lValueRequested)
-					m_touchedVariable[&n] = varDecl;
-		}
-		else if (auto funCall = dynamic_cast<FunctionCall const*>(&n))
-		{
-			if (FunctionDefinition const* funDef = SMTChecker::inlinedFunctionCallToDefinition(*funCall))
-				m_children[&n].push_back(funDef);
-		}
-		return true;
-	};
-	auto edgeFun = [&](ASTNode const& _parent, ASTNode const& _child)
-	{
-		if (m_touchedVariable.count(&_child) || m_children.count(&_child))
-			m_children[&_parent].push_back(&_child);
-	};
-
-	ASTReduce reducer(nodeFun, edgeFun);
-	_node.accept(reducer);
+	Declaration const* declaration = _identifier.annotation().referencedDeclaration;
+	solAssert(declaration, "");
+	if (VariableDeclaration const* varDecl = dynamic_cast<VariableDeclaration const*>(declaration))
+		if (_identifier.annotation().lValueRequested)
+			m_touchedVariables.insert(varDecl);
 }
 
-vector<VariableDeclaration const*> VariableUsage::touchedVariables(ASTNode const& _node) const
+void VariableUsage::endVisit(FunctionCall const& _funCall)
 {
-	if (!m_children.count(&_node) && !m_touchedVariable.count(&_node))
-		return {};
+	if (auto const& funDef = SMTChecker::inlinedFunctionCallToDefinition(_funCall))
+		if (find(m_functionPath.begin(), m_functionPath.end(), funDef) == m_functionPath.end())
```

Comment:
This ensures that this terminates if e.g. ``f`` calls ``g`` and ``g`` calls ``f`` back, right?

Reply:
Yep, that's the idea

Reply:
Hmmmm but maybe we should pass the entire call stack from the SMTChecker to VariableUsage to have the exact same behavior

Reply:
Ok done

Reply:
Ok done